# Ejercicios Prácticos Módulo VI

La entrega será la copia de este notebook resuelta con el nombre del participante en el nombre del archivo.
* Recuerde copiar este notebook y trabajar sobre la copia.

## Opción 1

Realizar los 3 siguientes ejercicios, documentar paso a paso como es el proceso, imprimir el accuraccy y las imagenes resultantes de cada uno de ellos.

### Ejercicio 1

Hacer un clasificador de perros y gatos, definimos las imagenes de 180x180x1 y el número de clases es igual a 2, usar una arquitectura tipo lenet (Accuracy debe ser mayor al 70% para uqe se tome como completo).

In [6]:
import os
import numpy as numpy
import keras
from keras import layers
from tensorflow import data as tf_data
import matplotlib.pyplot as plt
import tensorflow as tf

Con este comando descargaremos el dataset de las imagenes de Perros y Gatos que utilizaremos para entrenar nuestro modelo

In [ ]:
!curl -O https://download.microsoft.com/download/3/E/1/3E1C3F21-ECDB-4869-8368-6DEBA77B919F/kagglecatsanddogs_5340.zip

Debemos descomprimir el archivo del dataset.

In [ ]:
!unzip -q kagglecatsanddogs_5340.zip

In [ ]:
!ls PetImages

El dataset de las imagenes que se descargo puede ser que contenga imagenes dañadas, por lo tanto es importante realizar una limpieza de nuestro dataset.
En este caso revisaremos imagen por imagen para saber cuales pudieran estar dañadas y en este caso las eliminamos, ya que no nos serán útiles.

In [7]:
num_deleted = 0
for folder_name in ("Cat", "Dog"):
    folder_path = os.path.join("PetImages", folder_name)
    for fname in os.listdir(folder_path):
        fpath = os.path.join(folder_path, fname)
        try:
            fobj = open(fpath, "rb")
            is_jfif = b"JFIF" in fobj.peek(10)
        finally:
            fobj.close()
        
        if not is_jfif:
            num_deleted += 1
            # Borramos la imagen corrupta
            os.remove(fpath)

print(f"Imagenes eliminadas: {num_deleted}")

Imagenes eliminadas: 1578


El total de Imagenes eliminadas fue de: 1578, es por eso que es importante hacer una limpieza a nuestro dataset.

In [9]:
image_size = (180,180) # Las instrucciones del ejercicio nos solicita definir las imagenes de 180x180
batch_size = 128 # Definimos el tamaño de nuestro batch, este realmente puede ser cualquiera, pero se toma la sugerencia del profe de 128

train_ds, val_ds = keras.utils.image_dataset_from_directory(
    "PetImages", # Directorio donde se encuentran nuestras imagenes del dataset, keras toma cada carpeta que tengamos como una clase, en este caso "Cats" y "Dogs"
    validation_split=0.2, # Utilizaremos el 80% para entrenamiento y el 20% para validación
    subset="both", # Le indicamos a keras que esperamos recibir ambos sets train y validation
    seed=701, # Establecemos una semilla de aleatoreidad, puede ser cualquiera a mi me gusta la 701
    image_size=image_size, # Definimos el tamaño de las imagenes
    batch_size=batch_size, # Definimos el tamaño del batch
    color_mode="grayscale" # Como el ejercicio pide que sea 180x180x1 el x1 del final hace referencia a escalas grises, actualmente nuestras imagenes son RGB x3
)

Found 23422 files belonging to 2 classes.
Using 18738 files for training.
Using 4684 files for validation.


### Ejercicio 2

Crea una red que clasifique el dataset stanford_dogs usando el extractor del VGG16
1. Definir un modelo que tenga el VGG16 como entrada
2. Aplanar el vector de rasgos
3. Definir una o varias capas densas con 120 clases de salida
4. Entrenar el modelo

### Ejercicio 3

YOLO
* Hacer un dataset de 100 imágenes con 3 clases
* Entrenar el modelo usando el dataset ya sea con la implementación de keras o la de ultralytics, se recomienda ultralytics.
* Entrenar el modelo y mostrar los resultados.

## Opción 2

Realiza una implementación que resuleva una problemática del día a día utilizando algunos de los modelos vistos en clase.
Ejemplo:
* Detector de mascotas
* Clasificador de ropa
* Detector de vehículos para una cochera
* Etc

Te puedes apoyar de la IA para que te ayude con las partes de código que no entiendas, con la condición de que eres responsable por la entrega (si la IA se equivoca no es excusa para cambiar la clasificación)
Debes poner un planteamiento, el código bien documentando, así como los resultados con una imagen que tu hayas tomado.